app/browser/reducers/windowsReducer.js

```diff
@@ -245,7 +246,52 @@ const createWindow = (state, action) => {
   const toolbarUserInterfaceScale = getSetting(settings.TOOLBAR_UI_SCALE)
 
   setImmediate(() => {
-    const win = new BrowserWindow(Object.assign(windowProps, browserOpts, {disposition: frameOpts.disposition}))
+    // normally macOS will open immediately-created windows from fullscreen
+    // parent windows as fullscreen
+    // but since we are showing the window async, we will set the window
+    // fullscreen once it is ready to be shown
+    // (browserOpts.fullscreen may already be set when loading from saved state,
+    // so this just sets it for other scenarios)
+    if (isDarwin && parentWindow && parentWindow.isFullScreen()) {
```

Comment:
this logic (all of what is in this setImmediate) might be better to pull out to a discrete function.  This would help with unit testing and could help with readability. Something like `getCreateWindowArgs` (passing what's needed). Then whatever is returned is directly passed to the call `new BrowserWindow`

Reply:
Unit tests would then be a great way to capture the behavior (of setting fullscreen to false) and the tests themselves would act as documentation (if someone makes a change which breaks the test).  Could we get some unit tests for this?

Reply:
I have split all the window creation, showing, deferred showing, and timeout showing logic to a different place - _app/browser/windows.js_ which seemed an appropriate place which deals with calling functions on a muon BrowserWindow (and children). It's also the same place that the whenRendered action is handled, so that nicely puts all the logic in one place, including the temporary but ugly win.__blah properties - so they are at least encapsulated in one module.

I created tests for this new logic in _windowsTest.js_ handling:
- showing the window immediately
- showing the window if the timout expires
- showing the window on whenRendered firing for the window
- ensuring the window is fullscreen or maximized on each of those scenarios 

I also created tests for most of the existing action handler logic for APP_NEW_WINDOW in _windowReducerTest.js_, of which there were no existing tests, including tests for:
- creating the right tabs in the new window
- window position and size
- maximizing the window